XMonad/Layout/IndependentScreens.hs

```diff
@@ -186,3 +218,86 @@ unmarshallWindowSpace :: WindowSpace -> WindowSpace
 
 marshallWindowSpace s ws = ws { tag = marshall s  (tag ws) }
 unmarshallWindowSpace ws = ws { tag = unmarshallW (tag ws) }
+
+-- for marshallPP to work with XMonad.Actions.CopyWindow
+-- only ppHidden is changed
+marshallPP' :: ScreenId -> PP -> PP
+marshallPP' s pp = pp {
+    ppCurrent           = ppCurrent         pp . snd . unmarshall,
+    ppVisible           = ppVisible         pp . snd . unmarshall,
+    ppHidden            = ppHidden          pp ,
+    ppHiddenNoWindows   = ppHiddenNoWindows pp . snd . unmarshall,
+    ppUrgent            = ppUrgent          pp . snd . unmarshall,
+    ppSort              = fmap (marshallSort' s) (ppSort pp)
+    }
+
+-- if use original marshallSort, getSortByIndex will not work properly, since it sends the "video" instead of "0_video" back, and getSortByIndex has no way of knowing the order of "video" if the workspaces are declared with withScreens. However, getSortByTag works. 
+marshallSort' :: ScreenId -> ([WindowSpace] -> [WindowSpace]) -> ([WindowSpace] -> [WindowSpace])
+marshallSort' s vSort = pScreens . vSort . vScreens where
+        onScreen ws = unmarshallS (tag ws) == s
+        vScreens    = filter onScreen
+        pScreens    = id
+
+-- do something with the Nth workspace
+-- can be used with >9 workspaces
+-- to let withNthWorkspace in XMonad.Actions.DynamicWorkspaces be screen aware
+withNthWorkspace' :: (String -> WindowSet -> WindowSet) -> Int -> X ()
+withNthWorkspace' job wnum = do 
```

Comment:
Perhaps `withNthWorkspaceScreen` or something like that?  In general I think the `f'` idiom is only valid if there's exactly one way in which `f` may be changed, which is not the case here (nor is it for any of the functions below)

Reply:
My idea is that if someone use getSortByIndex together with dynamicWorkspace (another module), these function will kind of serve as an replacement of the "withNthWorkspace". Your suggestion make sense.

Reply:
Have renamed it.